conjure-java-core/src/main/java/com/palantir/conjure/java/types/UnionGenerator.java

```diff
@@ -577,17 +588,22 @@ private static TypeSpec generateBase(
                 .collect(Collectors.toList());
     }
 
-    private static AnnotationSpec wrapperConstructorParameterAnnotation(FieldDefinition field) {
+    private static AnnotationSpec wrapperConstructorParameterAnnotation(
+            FieldDefinition field, Map<com.palantir.conjure.spec.TypeName, TypeDefinition> typesMap) {
         AnnotationSpec.Builder builder = AnnotationSpec.builder(JsonSetter.class)
                 .addMember("value", "$S", field.getFieldName().get());
-        if (isListSetOrMap(field.getType())) {
+        Type dealiased = TypeFunctions.toConjureTypeWithoutAliases(field.getType(), typesMap);
+        if (supportsNullCoercion(dealiased)) {
             builder.addMember("nulls", "$T.AS_EMPTY", Nulls.class);
         }
         return builder.build();
     }
 
-    private static boolean isListSetOrMap(Type type) {
-        return type.accept(TypeVisitor.IS_LIST) || type.accept(TypeVisitor.IS_SET) || type.accept(TypeVisitor.IS_MAP);
+    private static boolean supportsNullCoercion(Type type) {
+        return type.accept(TypeVisitor.IS_LIST)
+                || type.accept(TypeVisitor.IS_SET)
+                || type.accept(TypeVisitor.IS_MAP)
+                || type.accept(TypeVisitor.IS_OPTIONAL);
```

Comment:
interestingly I think `DefaultableTypeVisitor` implements the same logic (true for all container types, false otherwise)

Reply:
good call, I think `DefaultableTypeVisitor` name makes sense since this is for types with a default value as well.